### PR TITLE
ARROW-5029: [C++] Fix compilation warnings in release mode

### PR DESCRIPTION
**Pull request by pitrou**
Also fix a severe memory leak in CUDA (CUDA buffers were never deallocated).

### CODE REVIEW

**Comment by pitrou** [comment]
@kou This might deserve doing another RC.

File: dev/release/verify-release-candidate.sh [hunk 1]
```diff
@@ -202,6 +202,7 @@ test_and_install_cpp() {
 
   ARROW_CMAKE_OPTIONS="
 ${ARROW_CMAKE_OPTIONS}
+-DARROW_DEPENDENCY_SOURCE=BUNDLED
```
Comment (pitrou) [inline dev/release/verify-release-candidate.sh]: Not sure it's ok to include this here.

Comment (bkietz) [reply dev/release/verify-release-candidate.sh]: What error do you get when using conda for dependencies?

Comment (pitrou) [reply dev/release/verify-release-candidate.sh]: When running the release script, I don't use conda for dependencies. AFAIU it should check downloading those.

File: cpp/src/arrow/io/compressed.cc [hunk 2]
```diff
@@ -46,7 +46,7 @@ class CompressedOutputStream::Impl {
   Impl(MemoryPool* pool, Codec* codec, const std::shared_ptr<OutputStream>& raw)
       : pool_(pool), raw_(raw), codec_(codec), is_open_(true), compressed_pos_(0) {}
 
-  ~Impl() { DCHECK(Close().ok()); }
+  ~Impl() { DCHECK_OK(Close()); }
```
Comment (pitrou) [inline cpp/src/arrow/io/compressed.cc]: These are other potential leaks.

File: dev/release/verify-release-candidate.sh (hunk 1 above)
Comment (pitrou) [reply dev/release/verify-release-candidate.sh]: @xhochy

**Comment by pitrou** [comment]
AppVeyor build:
https://ci.appveyor.com/project/pitrou/arrow/builds/23399732
Travis-CI build:
https://travis-ci.org/pitrou/arrow/builds/512073233

File: dev/release/verify-release-candidate.sh (hunk 1 above)
Comment (kszucs) [reply dev/release/verify-release-candidate.sh]: I think `bundled` is the closest one to the previous behaviour.

File: cpp/src/arrow/util/logging.h [hunk 3]
```diff
@@ -81,22 +81,38 @@ enum class ArrowLogLevel : int {
...
-#define DCHECK_GE(val1, val2) \
+#define DCHECK_GE(val1, val2)            \
+  while (false) ARROW_IGNORE_EXPR(val1); \
+  while (false) ARROW_IGNORE_EXPR(val2); \
   while (false) ::arrow::util::detail::NullLog()
-#define DCHECK_GT(val1, val2) \
+#define DCHECK_GT(val1, val2)            \
+  while (false) ARROW_IGNORE_EXPR(val1); \
+  while (false) ARROW_IGNORE_EXPR(val2); \
   while (false) ::arrow::util::detail::NullLog()
```
Comment (wesm) [inline cpp/src/arrow/util/logging.h]: I'm not super keen on these changes. I would rather suppress the compiler warnings

**Review by wesm — APPROVED** [review]
+1, with reservations noted. I'm going to check this out and take a look locally before merging

File: cpp/src/arrow/util/logging.h (hunk 3 above)
Comment (pitrou) [reply cpp/src/arrow/util/logging.h]: You mean unconditionally? I'm not sure it's possible to suppress them selectively in the macro...

File: dev/release/verify-release-candidate.sh (hunk 1 above)
Comment (kou) [reply dev/release/verify-release-candidate.sh]: I think that `-DARROW_DEPENDENCY_SOURCE=AUTO` is better.
Anyway, please remove this change from this pull request.
It's not related to "[C++] Fix compilation warnings in release mode". I'll create a pull request for this case.

Comment (pitrou) [reply dev/release/verify-release-candidate.sh]: Done.

**Comment by wesm** [comment]
I built on this branch without the changes to logging.h and things are OK:

https://gist.github.com/wesm/d1a6a6e2a0b1e9f082b9fe14a98502d5

Per my comments in https://github.com/apache/arrow/pull/4027, I think either we should

* Add `ARROW_IGNORE_EXPR(...)` statements to suppress warnings on a case-by-case basis rather than for every dcheck
* Add `-Wno-unused-variable` in release builds (the "nuke it from orbit" approach)
* Be OK with warnings associated with DCHECK statements in release builds 

If we have actual unused variables, they'll show up in debug builds (right?). Part of why I'm unsure or opposed to these changes is that I've looked at other projects like glog and Kudu, and others that use similar DCHECK statements (ours originated from Apache Kudu) and they haven't deemed it necessary to make the changes made here or in https://github.com/apache/arrow/pull/4027 (I seem to recall one or two other PRs where we debated this)

see e.g. glog https://github.com/google/glog/blob/master/src/glog/logging.h.in#L1034

The DCHECK concept stems from glog so I don't think we should deviate significantly from what they are doing.

**Comment by pitrou** [comment]
In this case the warnings helped uncover the memory leaks. So I don't think it's a good idea to silence them unconditionally.

I'm also unsure what we gain by silencing them one by one...

**Comment by pitrou** [comment]
If you look e.g. at `DLOG_IF(condition)`, the warning is silenced by the `(true || !(condition))` construct.

**Comment by wesm** [comment]
Alright. I'll merge things as is with my uncertainty on the record. I guess I'm paranoid about compilers that might be bad at dead code elimination and so there could be executable bloat. Most modern compilers (gcc / clang) should delete these parts of the AST but who knows what MSVC will do =)

**Comment by kou** [comment]
https://travis-ci.org/pitrou/arrow/builds/512225430 is green.
I'll merge this for RC4.

**Pull request closed without merging**